django_netjsonconfig/base/template.py

```diff
@@ -99,6 +138,36 @@ def clean(self, *args, **kwargs):
             self.auto_cert = False
         if self.type == 'vpn' and not self.config:
             self.config = self.vpn.auto_client(auto_cert=self.auto_cert)
+        if self.sharing == 'public' or self.sharing == 'secret_key':
+            if not self.description:
+                raise ValidationError({'description': _('Please enter a description of the shared template')})
+            self.config = self._update_config_with_variable(self.config, self.samples)
+        super(AbstractTemplate, self).clean(*args, **kwargs)
+
+    def _update_config_with_variable(self, config, variable):
```

Comment:
can you explain how this is being used right now?

Reply:
This function is used to transfer the content of `variable` to the exact key:values of `config`. that is it replaces the value of config with the corresponding value in `samples`
So assuming that `config` has an IP address that is a variable ( can be changed), all the admin need to do is specify the interface and the new IP address in the `samples` field and that will be copied to the config. 
this equally allows `config` to handle the validation, depending on the backend that is used.
The test case `test_sample_variable_replace_config_values` in `test_template.py` test for this

Reply:
wait a moment, it seems to me you are reimplementing something that is already implemented.

Study the base model here:
https://github.com/openwisp/django-netjsonconfig/blob/master/django_netjsonconfig/base/base.py

Look at `get_backend_instance`, `clean_netjsonconfig_backend` and `clean`.

Try to define a `get_context` method in the `Template` model which returns the sample values, this will pass the sample values as the [netjsonconfig context](http://netjsonconfig.openwisp.org/en/stable/general/basics.html#context), and netjsonconfig will take care of the validation.

But beware! The validation is done both in the admin web UI (with JS) and in the model validation (this way that I explained), so that after this PR is ready I suggest you to focus on solving the issue for the web UI with JS, is it clear? If not, please ask.